jit: arm: Cleaning up vasm-arm.

cmuellner opened the pull request:
* Remove lowering: opi -> ldimm;op
 * Add/fix missing/broken opi emit functions
 * Remove unsigned extensions before byte instructions
 * Add unsigned extensions after byte/word instructions if needed

Signed-off-by: Christoph Muellner <christoph.muellner@theobroma-systems.com>

cmuellner commented:
@mxw please review.
Note, that this PR is extracted from #7797.
This patch was only tested on top of PR #7744, #7788, #7795, #7805, #7806, #7807, #7808, and #7810. It will not work without one or more of these PRs applied for sure (i.e. you will get incorrect runtime behaviour resulting in triggered assertions and or aborts/segmentation faults).

A performance analysis (on top of the named PRs) showed a performance speedup of 1.5% on my arm64 machine (Ubuntu Xenial) in OSS-performance's wikimedia benchmark (comparing the average of 5 runs with cbc6a1fe02a).

jim-saxman commented:
@cmuellner I wasn't able to apply string of PR's cleanly ( #7744, #7788, #7795, #7805, #7806, #7807, #7808, and #7810. #7795 failed compilation. Skipping #7795 passes quick on CentOS.

cmuellner commented:
@jim-saxman thank's for having a look at that.

Currently my hhvm-arm64-next remote branch has all my PRs on top of master (cbc6a1f).
@swalk-cavium already confirmed that this branch compiles.

I created that remote branch by simply cherry-picking the commits of the PRs one by one (note, that two PRs consist of two commits).  I saw a problem only with the last (this) PR, but trivial to fix. If you share details (here, on the list or via email) what does not work, then I will have a look.

And regarding the compilation issues: please do a clean re-build (`git clean -dfx && rm -rf third-party && git submodule update --init --recursive && [...]`) as I haven't seen such issues during my tests (on Ubuntu and Fedora) and could not reproduce them just now (Fedora). If the problem persists feel free to share details, then I will have a look.

cmuellner commented:
@jim-saxman I can now also confirm that compilation (using branch hhvm-arm64-next) on CentOS 7 works and a run of the 'all' test set showed no regressions (comparing with the nightly result [1]).

[1] http://hhvm-arm64.theobroma-systems.com/1493258402683224003/test.txt

jim-saxman commented:
@cmuellner Thanks. I'll try again. I was able to confirm no regressions on 'all' and OSS.

jim-saxman commented:
@cmuellner  I think my compilation issue is that #7795 removed the subbi vasm and #7806 uses the subbi vasm.

cmuellner commented:
@jim-saxman sorry for not warning about that. As I said in #7795, as I was requested to prepare individual PRs for a commit series. So I prepared them independent from each other (as I cannot influence the order or merging). And I want my PRs to be as complete as possible (imagine #7795 gets nack'ed, then I suddenly want the fold functions for subb/subbi).

jim-saxman commented:
@cmuellner Sounds good to me. Dependencies sure are fun.

mxw commented:
@cmuellner, @jim-saxman—Yeah, sorry folks.  Dependency tracking on GitHub is nonexistent.  I'm planning to review these in the order in which you put them up, and land them one-by-one, so hopefully that will avoid bugs sneaking in.

cmuellner commented:
As there is a conflict after merging #7805 I'm going to rebase (and force-push) this.

cmuellner commented:
Note, that this PR currently depends on #7744, #7807, #7808 and #7833.

cmuellner commented:
I've added two things to this PR:

 * documentation about the early-truncation policy (based on my design proposal writeup on the mailing list)
 * Adapt movz{} and movt{} emitters according the policy

Benchmarks are currently running. I'll post the performance improvement (or decay) when they finish (I'm using OSS-performance's Mediawiki with 20 runs each and use the relative delta of the average).

cmuellner commented:
The accumulated sequence of PR #7744, #7807, #7808, #7833 and #7811 shows an improvement of 2.1%.

I've learned today, that OSS performance got some great improvements recently (I was using a 1-year old git-master version before), which reduce the jitter of the RPS results extremely (from >10%  to <1%). Besides that a lot of other stabilisation improvements and updates have been done. Therefore I've simplified my performance measurement methodology to make three runs of OSS performance (mediawiki) and compare the average.

cmuellner commented:
@swalk-cavium just pinged me via email, that there are self-`mov` instructions in the emitted code. The reason is, that optimizeCopies() works on the vasm instruction `copy{}`, but not on the emitted aarch64 code. I've decided to address this by only emitting the `mov` instruction if `i.s != i.d`. Note, that this might result in even better performance. I'll push as soon as the tests finish without regressions.

cmuellner commented:
The self-`mov`s @swalk-cavium observed could be addressed by putting the condition `i.s != i.d` to the emit function of the zero-extensions. However, that resulted in some obsolete non-self `mov`s being emitted. The underlying reason is, that obsolete `copy{}`s will only be eliminated on vasm level through `optimizeCopies()`. I decided therefore to replace the zero-extensions by `copy{}`s in the lowering pass and remove the emit functions.

Besides that I've rebased, merged #7744 into this PR (they were overlapping) and I've introduced the functions `bAsUb()` and `wAsUw()` to address #7807.

Testing showed no regressions on Ubuntu/Fedora (all set).
Performance test showed ~+1% RPS in Mediawiki.

mxw commented on hphp/runtime/vm/jit/vasm-arm.cpp:
```diff
@@ -1487,77 +1483,71 @@ void lower(const VLS& e, pushpm& i, Vlabel b, size_t z) {
+    lowerVptr(i.s1, v);
+    auto tmp0 = v.makeReg();
+    auto tmp1 = v.makeReg();
+    v << copy{i.s0, tmp0};
```
Why is this `copy{}` necessary?

cmuellner replied:
Because `cmpbm.s0` is a Vreg8 and `cmpl.s0` is a Vreg32 and zero-extension in the lowering is replaced by `copy{}`.
Same is true for `cmpwm.s0` (Vreg16). I.e. it does not compile without the `copy{}` or a zero-extension. And since `lowerForARM()` is called before `optimizeCopies()` the `copy{}` will be eliminated anyways.

mxw replied:
What do you mean by "it does not compile"?  What happens if you omit the copy, and just use the source directly?

cmuellner replied:
```
Scanning dependencies of target hphp_runtime_static
[ 52%] Building CXX object hphp/runtime/CMakeFiles/hphp_runtime_static.dir/vm/jit/vasm-arm.cpp.o
/mnt/cmuellner/hhvm-debug/hphp/runtime/vm/jit/vasm-arm.cpp: In instantiation of ‘HPHP::jit::{anonymous}::lower_cmpm(const HPHP::jit::VLS&, cmpm&, HPHP::jit::Vlabel, size_t)::<lambda(HPHP::jit::Vout&)> [with load_op = HPHP::jit::loadzbl; cmp_op = HPHP::jit::cmpl; cmpm = HPHP::jit::cmpbm]’:
/mnt/cmuellner/hhvm-debug/hphp/runtime/vm/jit/vasm-arm.cpp:1534:15:   required from ‘struct HPHP::jit::{anonymous}::lower_cmpm(const HPHP::jit::VLS&, cmpm&, HPHP::jit::Vlabel, size_t) [with load_op = HPHP::jit::loadzbl; cmp_op = HPHP::jit::cmpl; cmpm = HPHP::jit::cmpbm; size_t = long unsigned int]::<lambda(struct HPHP::jit::Vout&)>’
/mnt/cmuellner/hhvm-debug/hphp/runtime/vm/jit/vasm-arm.cpp:1533:13:   required from ‘void HPHP::jit::{anonymous}::lower_cmpm(const HPHP::jit::VLS&, cmpm&, HPHP::jit::Vlabel, size_t) [with load_op = HPHP::jit::loadzbl; cmp_op = HPHP::jit::cmpl; cmpm = HPHP::jit::cmpbm; size_t = long unsigned int]’
/mnt/cmuellner/hhvm-debug/hphp/runtime/vm/jit/vasm-arm.cpp:1542:39:   required from here
/mnt/cmuellner/hhvm-debug/hphp/runtime/vm/jit/vasm-arm.cpp:1537:7: error: could not convert ‘i.HPHP::jit::cmpbm::s0’ from ‘HPHP::jit::Vreg8 {aka HPHP::jit::Vr<HPHP::jit::Reg8>}’ to ‘HPHP::jit::Vreg32 {aka HPHP::jit::Vr<HPHP::jit::Reg32>}’
     v << cmp_op{i.s0, tmp1, i.sf};
       ^
/mnt/cmuellner/hhvm-debug/hphp/runtime/vm/jit/vasm-arm.cpp: In instantiation of ‘HPHP::jit::{anonymous}::lower_cmpm(const HPHP::jit::VLS&, cmpm&, HPHP::jit::Vlabel, size_t)::<lambda(HPHP::jit::Vout&)> [with load_op = HPHP::jit::loadw; cmp_op = HPHP::jit::cmpl; cmpm = HPHP::jit::cmpwm]’:
/mnt/cmuellner/hhvm-debug/hphp/runtime/vm/jit/vasm-arm.cpp:1534:15:   required from ‘struct HPHP::jit::{anonymous}::lower_cmpm(const HPHP::jit::VLS&, cmpm&, HPHP::jit::Vlabel, size_t) [with load_op = HPHP::jit::loadw; cmp_op = HPHP::jit::cmpl; cmpm = HPHP::jit::cmpwm; size_t = long unsigned int]::<lambda(struct HPHP::jit::Vout&)>’
/mnt/cmuellner/hhvm-debug/hphp/runtime/vm/jit/vasm-arm.cpp:1533:13:   required from ‘void HPHP::jit::{anonymous}::lower_cmpm(const HPHP::jit::VLS&, cmpm&, HPHP::jit::Vlabel, size_t) [with load_op = HPHP::jit::loadw; cmp_op = HPHP::jit::cmpl; cmpm = HPHP::jit::cmpwm; size_t = long unsigned int]’
/mnt/cmuellner/hhvm-debug/hphp/runtime/vm/jit/vasm-arm.cpp:1546:37:   required from here
/mnt/cmuellner/hhvm-debug/hphp/runtime/vm/jit/vasm-arm.cpp:1537:7: error: could not convert ‘i.HPHP::jit::cmpwm::s0’ from ‘HPHP::jit::Vreg16 {aka HPHP::jit::Vr<HPHP::jit::Reg16>}’ to ‘HPHP::jit::Vreg32 {aka HPHP::jit::Vr<HPHP::jit::Reg32>}’
/mnt/cmuellner/hhvm-debug/hphp/runtime/vm/jit/vasm-arm.cpp:1100:6: error: ‘void HPHP::jit::{anonymous}::lower_impl(HPHP::jit::Vunit&, HPHP::jit::Vlabel, size_t, Lower) [with Lower = HPHP::jit::{anonymous}::lower_cmpm(const HPHP::jit::VLS&, cmpm&, HPHP::jit::Vlabel, size_t) [with load_op = HPHP::jit::loadzbl; cmp_op = HPHP::jit::cmpl; cmpm = HPHP::jit::cmpbm; size_t = long unsigned int]::<lambda(HPHP::jit::Vout&)>; size_t = long unsigned int]’, declared using local type ‘HPHP::jit::{anonymous}::lower_cmpm(const HPHP::jit::VLS&, cmpm&, HPHP::jit::Vlabel, size_t) [with load_op = HPHP::jit::loadzbl; cmp_op = HPHP::jit::cmpl; cmpm = HPHP::jit::cmpbm; size_t = long unsigned int]::<lambda(HPHP::jit::Vout&)>’, is used but never defined [-fpermissive]
 void lower_impl(Vunit& unit, Vlabel b, size_t i, Lower lower) {
      ^
/mnt/cmuellner/hhvm-debug/hphp/runtime/vm/jit/vasm-arm.cpp:1100:6: error: ‘void HPHP::jit::{anonymous}::lower_impl(HPHP::jit::Vunit&, HPHP::jit::Vlabel, size_t, Lower) [with Lower = HPHP::jit::{anonymous}::lower_cmpm(const HPHP::jit::VLS&, cmpm&, HPHP::jit::Vlabel, size_t) [with load_op = HPHP::jit::loadw; cmp_op = HPHP::jit::cmpl; cmpm = HPHP::jit::cmpwm; size_t = long unsigned int]::<lambda(HPHP::jit::Vout&)>; size_t = long unsigned int]’, declared using local type ‘HPHP::jit::{anonymous}::lower_cmpm(const HPHP::jit::VLS&, cmpm&, HPHP::jit::Vlabel, size_t) [with load_op = HPHP::jit::loadw; cmp_op = HPHP::jit::cmpl; cmpm = HPHP::jit::cmpwm; size_t = long unsigned int]::<lambda(HPHP::jit::Vout&)>’, is used but never defined [-fpermissive]
hphp/runtime/CMakeFiles/hphp_runtime_static.dir/build.make:9572: recipe for target 'hphp/runtime/CMakeFiles/hphp_runtime_static.dir/vm/jit/vasm-arm.cpp.o' failed
make[2]: *** [hphp/runtime/CMakeFiles/hphp_runtime_static.dir/vm/jit/vasm-arm.cpp.o] Error 1
CMakeFiles/Makefile2:1725: recipe for target 'hphp/runtime/CMakeFiles/hphp_runtime_static.dir/all' failed
make[1]: *** [hphp/runtime/CMakeFiles/hphp_runtime_static.dir/all] Error 2
Makefile:127: recipe for target 'all' failed
make: *** [all] Error 2
```

mxw replied:
Ah, because the `Vr` specializations don't interconvert; you have to proxy through `Vreg` itself.  So just nix the `copy{}`, and do `Vreg(i.s0)`, and that should work.

cmuellner replied:
The `copy{}`-pattern is used in the functions `lower(movtdb& i)` and `lower(movtdq& i)` above as well for the same reason (in the already existing code). In general the `copy{}` is not harmful, since `optimizeCopies()` cleans up later anyways (you proposed to emit a `copy{}` in `simplify_andi()` so I assume, that relying upon optimizeCopies() is ok).

When doing the proposed Vreg-cast/conversion/proxy trick I get:

```
[ 51%] Building CXX object hphp/runtime/CMakeFiles/hphp_runtime_static.dir/vm/jit/vasm-arm.cpp.o
/mnt/cmuellner/hhvm/hphp/runtime/vm/jit/vasm-arm.cpp: In instantiation of ‘HPHP::jit::{anonymous}::lower_cmpm(const HPHP::jit::VLS&, cmpm&, HPHP::jit::Vlabel, size_t)::<lambda(HPHP::jit::Vout&)> [with load_op = HPHP::jit::loadzbl; cmp_op = HPHP::jit::cmpl; cmpm = HPHP::jit::cmpbm]’:
/mnt/cmuellner/hhvm/hphp/runtime/vm/jit/vasm-arm.cpp:1534:15:   required from ‘struct HPHP::jit::{anonymous}::lower_cmpm(const HPHP::jit::VLS&, cmpm&, HPHP::jit::Vlabel, size_t) [with load_op = HPHP::jit::loadzbl; cmp_op = HPHP::jit::cmpl; cmpm = HPHP::jit::cmpbm; size_t = long unsigned int]::<lambda(struct HPHP::jit::Vout&)>’
/mnt/cmuellner/hhvm/hphp/runtime/vm/jit/vasm-arm.cpp:1533:13:   required from ‘void HPHP::jit::{anonymous}::lower_cmpm(const HPHP::jit::VLS&, cmpm&, HPHP::jit::Vlabel, size_t) [with load_op = HPHP::jit::loadzbl; cmp_op = HPHP::jit::cmpl; cmpm = HPHP::jit::cmpbm; size_t = long unsigned int]’
/mnt/cmuellner/hhvm/hphp/runtime/vm/jit/vasm-arm.cpp:1542:39:   required from here
/mnt/cmuellner/hhvm/hphp/runtime/vm/jit/vasm-arm.cpp:1537:17: error: call of overloaded ‘Vreg(HPHP::jit::Vreg8&)’ is ambiguous
     v << cmp_op{Vreg(i.s0), tmp, i.sf};
                 ^~~~~~~~~~
In file included from /mnt/cmuellner/hhvm/hphp/runtime/vm/jit/arg-group.h:21:0,
                 from /mnt/cmuellner/hhvm/hphp/runtime/vm/jit/vasm-instr.h:21,
                 from /mnt/cmuellner/hhvm/hphp/runtime/vm/jit/vasm-unit.h:26,
                 from /mnt/cmuellner/hhvm/hphp/runtime/vm/jit/vasm-gen.h:24,
                 from /mnt/cmuellner/hhvm/hphp/runtime/vm/jit/vasm-arm.cpp:72:
/mnt/cmuellner/hhvm/hphp/runtime/vm/jit/vasm-reg.h:59:18: note: candidate: HPHP::jit::Vreg::Vreg(HPHP::jit::Reg8)
   /* implicit */ Vreg(Reg8 r)   : rn(int(r)) {}
                  ^~~~
/mnt/cmuellner/hhvm/hphp/runtime/vm/jit/vasm-reg.h:55:12: note: candidate: HPHP::jit::Vreg::Vreg(size_t)
   explicit Vreg(size_t r) : rn(r) {}
            ^~~~
/mnt/cmuellner/hhvm/hphp/runtime/vm/jit/vasm-reg.h:41:8: note: candidate: constexpr HPHP::jit::Vreg::Vreg(const HPHP::jit::Vreg&)
 struct Vreg {
        ^~~~
/mnt/cmuellner/hhvm/hphp/runtime/vm/jit/vasm-reg.h:41:8: note: candidate: constexpr HPHP::jit::Vreg::Vreg(HPHP::jit::Vreg&&)
/mnt/cmuellner/hhvm/hphp/runtime/vm/jit/vasm-arm.cpp: In instantiation of ‘HPHP::jit::{anonymous}::lower_cmpm(const HPHP::jit::VLS&, cmpm&, HPHP::jit::Vlabel, size_t)::<lambda(HPHP::jit::Vout&)> [with load_op = HPHP::jit::loadw; cmp_op = HPHP::jit::cmpl; cmpm = HPHP::jit::cmpwm]’:
/mnt/cmuellner/hhvm/hphp/runtime/vm/jit/vasm-arm.cpp:1534:15:   required from ‘struct HPHP::jit::{anonymous}::lower_cmpm(const HPHP::jit::VLS&, cmpm&, HPHP::jit::Vlabel, size_t) [with load_op = HPHP::jit::loadw; cmp_op = HPHP::jit::cmpl; cmpm = HPHP::jit::cmpwm; size_t = long unsigned int]::<lambda(struct HPHP::jit::Vout&)>’
/mnt/cmuellner/hhvm/hphp/runtime/vm/jit/vasm-arm.cpp:1533:13:   required from ‘void HPHP::jit::{anonymous}::lower_cmpm(const HPHP::jit::VLS&, cmpm&, HPHP::jit::Vlabel, size_t) [with load_op = HPHP::jit::loadw; cmp_op = HPHP::jit::cmpl; cmpm = HPHP::jit::cmpwm; size_t = long unsigned int]’
/mnt/cmuellner/hhvm/hphp/runtime/vm/jit/vasm-arm.cpp:1546:37:   required from here
/mnt/cmuellner/hhvm/hphp/runtime/vm/jit/vasm-arm.cpp:1537:17: error: call of overloaded ‘Vreg(HPHP::jit::Vreg16&)’ is ambiguous
     v << cmp_op{Vreg(i.s0), tmp, i.sf};
                 ^~~~~~~~~~
In file included from /mnt/cmuellner/hhvm/hphp/runtime/vm/jit/arg-group.h:21:0,
                 from /mnt/cmuellner/hhvm/hphp/runtime/vm/jit/vasm-instr.h:21,
                 from /mnt/cmuellner/hhvm/hphp/runtime/vm/jit/vasm-unit.h:26,
                 from /mnt/cmuellner/hhvm/hphp/runtime/vm/jit/vasm-gen.h:24,
                 from /mnt/cmuellner/hhvm/hphp/runtime/vm/jit/vasm-arm.cpp:72:
/mnt/cmuellner/hhvm/hphp/runtime/vm/jit/vasm-reg.h:58:18: note: candidate: HPHP::jit::Vreg::Vreg(HPHP::jit::Reg16)
   /* implicit */ Vreg(Reg16 r)  : rn(int(r)) {}
                  ^~~~
/mnt/cmuellner/hhvm/hphp/runtime/vm/jit/vasm-reg.h:55:12: note: candidate: HPHP::jit::Vreg::Vreg(size_t)
   explicit Vreg(size_t r) : rn(r) {}
            ^~~~
/mnt/cmuellner/hhvm/hphp/runtime/vm/jit/vasm-reg.h:41:8: note: candidate: constexpr HPHP::jit::Vreg::Vreg(const HPHP::jit::Vreg&)
 struct Vreg {
        ^~~~
/mnt/cmuellner/hhvm/hphp/runtime/vm/jit/vasm-reg.h:41:8: note: candidate: constexpr HPHP::jit::Vreg::Vreg(HPHP::jit::Vreg&&)
/mnt/cmuellner/hhvm/hphp/runtime/vm/jit/vasm-arm.cpp:1100:6: error: ‘void HPHP::jit::{anonymous}::lower_impl(HPHP::jit::Vunit&, HPHP::jit::Vlabel, size_t, Lower) [with Lower = HPHP::jit::{anonymous}::lower_cmpm(const HPHP::jit::VLS&, cmpm&, HPHP::jit::Vlabel, size_t) [with load_op = HPHP::jit::loadzbl; cmp_op = HPHP::jit::cmpl; cmpm = HPHP::jit::cmpbm; size_t = long unsigned int]::<lambda(HPHP::jit::Vout&)>; size_t = long unsigned int]’, declared using local type ‘HPHP::jit::{anonymous}::lower_cmpm(const HPHP::jit::VLS&, cmpm&, HPHP::jit::Vlabel, size_t) [with load_op = HPHP::jit::loadzbl; cmp_op = HPHP::jit::cmpl; cmpm = HPHP::jit::cmpbm; size_t = long unsigned int]::<lambda(HPHP::jit::Vout&)>’, is used but never defined [-fpermissive]
 void lower_impl(Vunit& unit, Vlabel b, size_t i, Lower lower) {
      ^~~~~~~~~~
/mnt/cmuellner/hhvm/hphp/runtime/vm/jit/vasm-arm.cpp:1100:6: error: ‘void HPHP::jit::{anonymous}::lower_impl(HPHP::jit::Vunit&, HPHP::jit::Vlabel, size_t, Lower) [with Lower = HPHP::jit::{anonymous}::lower_cmpm(const HPHP::jit::VLS&, cmpm&, HPHP::jit::Vlabel, size_t) [with load_op = HPHP::jit::loadw; cmp_op = HPHP::jit::cmpl; cmpm = HPHP::jit::cmpwm; size_t = long unsigned int]::<lambda(HPHP::jit::Vout&)>; size_t = long unsigned int]’, declared using local type ‘HPHP::jit::{anonymous}::lower_cmpm(const HPHP::jit::VLS&, cmpm&, HPHP::jit::Vlabel, size_t) [with load_op = HPHP::jit::loadw; cmp_op = HPHP::jit::cmpl; cmpm = HPHP::jit::cmpwm; size_t = long unsigned int]::<lambda(HPHP::jit::Vout&)>’, is used but never defined [-fpermissive]
/mnt/cmuellner/hhvm/hphp/runtime/vm/jit/vasm-arm.cpp:1100:6: warning: ‘void HPHP::jit::{anonymous}::lower_impl(HPHP::jit::Vunit&, HPHP::jit::Vlabel, size_t, Lower) [with Lower = HPHP::jit::{anonymous}::lower_cmpm(const HPHP::jit::VLS&, cmpm&, HPHP::jit::Vlabel, size_t) [with load_op = HPHP::jit::loadw; cmp_op = HPHP::jit::cmpl; cmpm = HPHP::jit::cmpwm]::<lambda(HPHP::jit::Vout&)>]’ used but never defined
/mnt/cmuellner/hhvm/hphp/runtime/vm/jit/vasm-arm.cpp:1100:6: warning: ‘void HPHP::jit::{anonymous}::lower_impl(HPHP::jit::Vunit&, HPHP::jit::Vlabel, size_t, Lower) [with Lower = HPHP::jit::{anonymous}::lower_cmpm(const HPHP::jit::VLS&, cmpm&, HPHP::jit::Vlabel, size_t) [with load_op = HPHP::jit::loadzbl; cmp_op = HPHP::jit::cmpl; cmpm = HPHP::jit::cmpbm]::<lambda(HPHP::jit::Vout&)>]’ used but never defined
hphp/runtime/CMakeFiles/hphp_runtime_static.dir/build.make:11084: recipe for target 'hphp/runtime/CMakeFiles/hphp_runtime_static.dir/vm/jit/vasm-arm.cpp.o' failed
```

When grepping the HHVM source for "Vreg(" to find an example that does something similar, I was not successful to find anything.

mxw replied:
Oops, sorry for the delay; I thought I already commented here.

I suppose you'd have to do something like `Vreg r = i.s0;` and then use `r`, but the `copy{}` is also fine if it's what we're already doing elsewhere, I suppose (though I think avoiding it would be better in general, because it's unnecessary).

mxw commented on hphp/runtime/vm/jit/vasm-arm.cpp:
```diff
@@ -1487,77 +1483,71 @@ void lower(const VLS& e, pushpm& i, Vlabel b, size_t z) {
 }
 
+template<typename load_op, typename cmp_op, typename cmpm>
+void lower_cmpm(const VLS&e, cmpm& i, Vlabel b, size_t z) {
```
nit: `const VLS& e`

mxw commented on hphp/runtime/vm/jit/vasm-arm.cpp:
```diff
@@ -1197,45 +1232,6 @@ Y(storew, m)
-Y(cmpqi, cmpq, ldimmq, ISAR, Immed64(value), i.s1)
-Y(orqi, orq, ldimmq, ISLG(64), Immed64(value), U2(i.s1, i.d))
-Y(subli, subl, ldimml, ISAR, i.s0, U2(i.s1, i.d))
-Y(subqi, subq, ldimmq, ISAR, Immed64(value), U2(i.s1, i.d))
```
Hmm... what was the point of these if vixl can already handle all of these inputs?  I think this might have been implemented by lpathy, so the answer might be lost now.

cmuellner replied:
My guess: There are different ways to handle the fact that not all vasm instructions "are existing" on aarch64. E.g. `cmpb{}` does not exist. So one way is to lower it to `cmpl{}` and have no emit function for `cmpb{}`. That is also true for all these instructions (e.g. splitting up cmpqi{} if the immediate is to big). However, this approach (lower everything what cannot be directly emitted) is not used throughout the whole code. Once I gave it a try to enforce this at least for `ldimmN{}` (see https://github.com/cmuellner/hhvm/commit/4ba6b5756555852bfba45a2dec1638e2f3d4bf06), but that ended up in having to touch the register allocator, which assumes `ldimmb{}` to be ok to emit (the register allocator is running after lowerForArm()).

In general I like the idea of lowering everything what cannot be translated to aarch64 directly, but I am not sure if it is worth it given that a) vixl can handle that anyways (we would have to reimplement vixl-features with the risk of making it worse) and b) non-arm-specific code might need to be aware that it cannot use some vasm instructions (e.g. the register allocator). Therefore I prefer to only lower vasm instructions, which can be lowered or need to be lowered (for some reason). But I am open here to change my mind. :)

mxw replied:
I basically agree with your conclusion at the end—I think we should go through vasm-lower for specifically those instructions which need to be lowered (e.g., because they would otherwise require scratch registers) or because doing so is an optimization (e.g., the `movtq*{}` instructions, which otherwise copy-propagation couldn't eliminate).  I definitely don't think the distinction should be tied to "what exists" on the architecture.

mxw commented:
Sorry this took so long to land—I was working out of a different office and then on PTO.

Thanks for driving this through to the end, @cmuellner.  It's done a lot to clarify the assumptions and invariants around how we're filling registers.
